web/src/games/soupofletters/game.ts

```diff
@@ -0,0 +1,138 @@
+
+import { MAX_WORDS_IN_GAME, MAX_WORD_LEN, DRAW_AFTER_N_TIMERS, validOrientations, globalWordList } from './constants';
+import { newPuzzle, solvePuzzel, printPuzzle } from './puzzel'; 
+
+
+export interface ISingleLetter {
+    x: number;
+    y: number;
+    letter: string;
+}
+
+export interface ISolvedWord{
+    x:number;
+    y:number;
+    orientation: string;
+    word: string;
+    overlap: number;
+    solvedBy?: string;
+    letters: Array<ISingleLetter>;
+}
+
+export interface IG {
+    puzzel: Array<Array<string>>;
+    solution: Array<ISolvedWord>;
+    timeRef: number;
+    countTimerFired: number;
+}
+
+function isGameOver(G, ctx) {
+    // if timer fired too many times, end game 
+    if (G.countTimerFired >= (ctx.numPlayers * DRAW_AFTER_N_TIMERS )) { return true; } 
+    // if all words are found then true 
+    let isOver = true;
+    for (const s of G.solution) {
+        if (s.solvedBy === undefined) {
+            isOver = false;
+        }
+    }
+    return isOver;
+}
+
+export function isVictory(G, ctx) { 
+    // count the score of each player 
+    const playerScores = {}
+    for (const s of G.solution) {
+        if (s.solvedBy) {
+            playerScores[s.solvedBy] = playerScores[s.solvedBy] || 0; 
+            playerScores[s.solvedBy] = playerScores[s.solvedBy] + 1; 
+        }
+    }
+    // determine max score 
+    let scores = Object.keys( playerScores ).map(function ( key ) { return playerScores[key]; });
+    var maxScore = Math.max.apply( null, scores );
+    // determine the player(s) who got maxScore
+    let playerWithMaxScore = undefined; 
+    let numPlayersAtMaxScore = 0; 
+    for (var player in playerScores){
+        if (maxScore === playerScores[player]){
+            playerWithMaxScore = player;
+            numPlayersAtMaxScore = numPlayersAtMaxScore + 1;
+        }
+    }
+    // if the max-score is same for any two players, then declare draw
+    if (numPlayersAtMaxScore === 1){
+        return { winner: playerWithMaxScore};
+    } else if (numPlayersAtMaxScore > 1){
+        return { draw: true };
+    } else {
+        return;
+    }
+}
+
+function initialSetup(ctx, wordList=globalWordList, allowedOrientations=validOrientations) {
+    // determine the number of words in game 
+    let wordPerPerson = Math.floor(MAX_WORDS_IN_GAME/ctx.numPlayers);
+    let totalAllowedWords = wordPerPerson * ctx.numPlayers;
+    // select a fixed amount of random words from the list of words 
+    let randWords = [];
+    for (let i = 0; i < wordList.length; i++){
+        const rWord = wordList[Math.floor(Math.random()*wordList.length)].toUpperCase();
+        if ((rWord.length <= MAX_WORD_LEN) && !randWords.includes(rWord)) { randWords.push(rWord); }
+        if (randWords.length >= totalAllowedWords) { break; }
+    }
+    // create a puzzel, and if all words are not found then reduce words and try again
+    let puzzel;
```

Comment:
Typo: puzzle?

Reply:
I guess I am not like other Indians who win the spelling bee